src/ClimaLSM.jl

```diff
@@ -1,5 +1,164 @@
+"""
+auxiliary_vars(m::AbstractModel) = ()
+
+"""
+    make_rhs(model::AbstractModel)
+
+Return a `rhs!` function that updates state variables.
+
+`rhs!` should be compatible with OrdinaryDiffEq.jl solvers.
+"""
+function make_rhs(model::AbstractModel)
+    function rhs!(dY, Y, p, t) end
+    return rhs!
+end
+
+"""
+    make_update_aux(model::AbstractModel)
+
+Return an `update_aux!` function that updates auxiliary parameters `p`.
+"""
+function make_update_aux(model::AbstractModel)
+    function update_aux!(p, Y, t) end
+    return update_aux!
+end
+
+"""
+    make_ode_function(model::AbstractModel)
+
+Returns an `ode_function` that updates auxiliary variables and
+updates the prognostic state.
+
+`ode_function!` should be compatible with OrdinaryDiffEq.jl solvers.
+"""
+function make_ode_function(model::AbstractModel)
+    rhs! = make_rhs(model)
+    update_aux! = make_update_aux(model)
+    function ode_function!(dY, Y, p, t)
+        update_aux!(p, Y, t)
+        rhs!(dY, Y, p, t)
+    end
+    return ode_function!
+end
+
+"""
+    initialize_prognostic(model::AbstractModel,_)
+
+Returns a FieldVector of prognostic variables for `model` with the required
+structure.  This is a stub. We will change it, if possible,
+ as we determine what components require.
+An idea is to use something like `similar(coords)`.
+"""
+function initialize_prognostic(model::AbstractModel, _)
+    return ClimaCore.Fields.FieldVector(; model.model_name => [])
+end
+
+
+"""
+    initialize_auxiliary(model::AbstractModel,_)
+
+Returns a FieldVector of auxiliary variables for `model` with the required
+structure.  This is a stub. We will change it, if possible,
+ as we determine what components require.
+An idea is to use something like `similar(coords)`."""
+function initialize_auxiliary(model::AbstractModel, _)
+    return ClimaCore.Fields.FieldVector(; model.model_name => [])
+end
+
+"""
+    initialize_states(model::AbstractModel)
+
+Creates the prognostic and auxiliary states structures, sets all values
+to zero; returns the coordinates for the `model` domain.
+"""
+function initialize_states(model::AbstractModel) end
+
+#### LandModel Specific
+
+"""
+    struct LandModel{FT, sm <: AbstractModel{FT}, rm <: AbstractModel{FT}} <: AbstractModel{FT}
+
+A concrete type of `AbstractModel` for use in land surface modeling. Each component model of the
+`LandModel` is itself an `AbstractModel`.
+"""
+struct LandModel{FT, sm <: AbstractModel{FT}, rm <: AbstractModel{FT}} <:
+       AbstractModel{FT}
+    soil::sm
+    roots::rm
+end
+
+function initialize_states(land::LandModel)
+    Y_soil, p_soil, coords_soil = initialize_states(land.soil)
+    Y_roots, p_roots, coords_roots = initialize_states(land.roots)
+    Y = ClimaCore.Fields.FieldVector(;
+        soil = Y_soil.soil,
+        roots = Y_roots.roots,
+    )
+    p = ClimaCore.Fields.FieldVector(;
+        soil = p_soil.soil,
+        roots = p_roots.roots,
+    )
+    coords =
```

Comment:
not sure if we really need coords, but it might be useful to pass them back to the user so they have them handy for plotting, e.g.. These would be (x,y,z) locations (z=z_srf for surface variables) or (z) alone - column mode.

Reply:
Is there a better place to return `coords`? Not sure that returning them is necessarily a responsibility of `initialize_states`.

Reply:
Agreed that this might not be the right place..but I kept running into the block of computing the space 1x. We could split coordinates out like:
```
coordinates = compute_coordinates(soil) # computes the space first
Y, p = initialize_states(soil, coordinates) # or else store the coordinates or space in soil rather than the domain
```
But then for the LSM, we would have to do
```
coordinates = compute_coordinates(land) # what form would this take given distinct domains for each component?
Y,p = initialize_states(land, coordinates)
```

or
```
function initialize_states(land)
    coordinates_soil = compute_coordinates(land.soil)
    Y,p = initialize_states(land.soil, coordinates_soil)
    #etc
end
```
but maybe there is a better way?

Reply:
https://github.com/CliMA/ClimaLSM.jl/issues/3 to be revisited once more is in place. what we have is a place to start from.